dsgrid/dataset/growth_rates.py

```diff
@@ -0,0 +1,92 @@
+import logging
+
+import pyspark.sql.functions as F
+from pyspark.sql.types import IntegerType
+
+from dsgrid.dimension.base_models import DimensionType
+from dsgrid.exceptions import DSGInvalidQuery
+from dsgrid.query.models import ExponentialGrowthDatasetModel
+from dsgrid.utils.spark import get_unique_values
+
+
+logger = logging.getLogger(__name__)
+
+
+def apply_growth_rate_123(
+    dataset: ExponentialGrowthDatasetModel,
+    initial_value_df,
+    growth_rate_df,
+    time_columns,
+    model_year_column,
+    pivoted_columns,
+):
+    """Applies the growth rate to the initial_value dataframe.
+
+    Parameters
+    ----------
+    initial_value_df : pyspark.sql.DataFrame
+    growth_rate_df : pyspark.sql.DataFrame
+    time_columns : set[str]
+    model_year_column : str
+    pivoted_columns : set[str]
+
+    Returns
+    -------
+    pyspark.sql.DataFrame
+
+    """
+    initial_value_df, base_year = _check_model_years(
+        dataset, initial_value_df, growth_rate_df, model_year_column
+    )
+    gr_df = growth_rate_df
+    for column in pivoted_columns:
+        gr_col = column + "__gr"
+        gr_df = gr_df.withColumn(
+            gr_col,
+            F.pow((1 + F.col(column)), F.col(model_year_column).cast(IntegerType()) - base_year),
+        ).drop(column)
+
+    dim_columns = set(initial_value_df.columns) - pivoted_columns - time_columns
+    # TODO DT: data_source needs some thought. They are different in these two dfs.
```

Comment:
Per team discussion today, we will leave this as-is. Meghan and Lixi want this data_source value to change to either `derived` or a concatenation of the two data sources. That presents some complications for registering the resulting derived dataset. We will leave that open for now. #185